File: 1652-defuse-the-bomb/1652-defuse-the-bomb.py
class Solution:
    def decrypt(self, code: List[int], k: int) -> List[int]:
        n = len(code)
        ans = [0] * n
        
        for i in range(n):
            curr = 0
            if k > 0:
                for di in range(1, k + 1):
                    curr += code[(i + di) % n]
            else:
                for di in range(1, abs(k) + 1):
                    curr += code[(i - di + n) % n]
            ans[i] = curr
        
        return ans

File: 1652-defuse-the-bomb/1652-defuse-the-bomb.cpp
class Solution {
public:
    vector<int> decrypt(vector<int>& code, int k) {
        int n = code.size();
        vector<int> ans(n);
        
        // Define the initial window and initial sum
        int start = 1, end = k;
        long currSum = 0;
        
        // If k < 0, the starting point will be the end of the array.
        if (k < 0){
            start = n - abs(k);
            end = n - 1;
        }
        
        for (int i = start; i <= end; ++i) currSum += code[i];
        
        // Scan through the code array as i moving to the right, update the window sum.
        for (int i = 0; i < n; ++i){
            ans[i] = currSum;
            currSum = currSum - code[start % n] + code[(end + 1) % n];
            ++start;
            ++end;
        }
        
        return ans;
    }
};